src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -4508,5 +4510,206 @@ public void GetRelatedDocumentsDoesNotReturnOtherTypesOfDocuments()
 
             Assert.Single(solution.GetRelatedDocumentIds(regularDocumentId));
         }
+
+        [Fact]
+        public async Task TestFrozenPartialSolution1()
+        {
+            using var workspace = WorkspaceTestUtilities.CreateWorkspaceWithPartialSemantics();
+            var project = workspace.CurrentSolution.AddProject("CSharpProject", "CSharpProject", LanguageNames.CSharp);
+            project = project.AddDocument("Extra.cs", SourceText.From("class Extra { }")).Project;
+
+            // Because we froze before ever even looking at anything semantics related, we should have no documents in
+            // this project.
+            var frozenSolution = await project.Solution.WithFrozenPartialCompilationsAsync(CancellationToken.None);
+            var frozenProject = frozenSolution.Projects.Single();
+            Assert.Empty(frozenProject.Documents);
+
+            var frozenCompilation = await frozenProject.GetCompilationAsync();
+            Assert.Empty(frozenCompilation.SyntaxTrees);
+        }
+
+        [Fact]
+        public async Task TestFrozenPartialSolution2()
+        {
+            using var workspace = WorkspaceTestUtilities.CreateWorkspaceWithPartialSemantics();
+            var project = workspace.CurrentSolution.AddProject("CSharpProject", "CSharpProject", LanguageNames.CSharp);
+            project = project.AddDocument("Extra.cs", SourceText.From("class Extra { }")).Project;
+
+            await project.GetCompilationAsync();
+
+            // Because we froze after looking at anything semantics related, we should have the documents in this
+            // project.
+            var frozenSolution = await project.Solution.WithFrozenPartialCompilationsAsync(CancellationToken.None);
+            var frozenProject = frozenSolution.Projects.Single();
+            Assert.Single(frozenProject.Documents);
+
+            var frozenCompilation = await frozenProject.GetCompilationAsync();
+            Assert.Single(frozenCompilation.SyntaxTrees);
+            Assert.True(frozenCompilation.ContainsSyntaxTree(await frozenProject.Documents.Single().GetSyntaxTreeAsync()));
+        }
+
+        [Fact]
+        public async Task TestFrozenPartialSolution3()
```

Comment:
I wish these tests had more descriptive names than making me reverse-engineer it from the code. :smile: